java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -139,9 +141,37 @@ private static Tree skipParentheses(Tree parent) {
     return skip;
   }
 
-  private static boolean isSubtype(Type expression, Type target) {
-    return expression.isSubtypeOf(target);
+  private static boolean isUnnecessarySubtypeCast(Type expressionType, TypeCastTree typeCastTree, Type target) {
+    return expressionType.isSubtypeOf(target) && !isLambdaCastNeeded(expressionType, typeCastTree, target);
   }
+
+  private static boolean isLambdaCastNeeded(Type expressionType, TypeCastTree typeCastTree, Type target) {
+    ExpressionTree expression = ExpressionUtils.skipParentheses(typeCastTree.expression());
+    if (expression.is(Tree.Kind.LAMBDA_EXPRESSION)) {
+      List<MethodJavaSymbol> childMethods = getMethodSymbolsOf(expressionType);
+      List<MethodJavaSymbol> parentMethods = getMethodSymbolsOf(target);
+
+      Predicate<JavaSymbol> isDefaultMethod = JavaSymbol::isDefault;
+      Set<MethodJavaSymbol> childDefaultMethods = childMethods.stream().filter(isDefaultMethod).collect(Collectors.toSet());
+      Optional<MethodJavaSymbol> childAbstractMethod = childMethods.stream().filter(isDefaultMethod.negate()).findAny();
+      Set<MethodJavaSymbol> parentAbstractMethods = parentMethods.stream().filter(isDefaultMethod.negate()).collect(Collectors.toSet());
```

Comment:
I may be wrong but I tend to think that you don't actually need to collect that one but rather anyMatch it with a predicate you would compute from the result of the previous optional. 

so you should end up with a 
```

return parentAbstractMethods.stream.filter().anyMatch(predicate)

Reply:
It doesn't need to be collected all the time, but it needs to be collected in the following `else` branch, as the `parentAbstractMethodsSet::contains` may be called multiple times, so it cannot be replaced by a `anyMatch` in that case. I've updated the code accordingly, so that streams are not collected when not needed